File: py-src/services/slack_service.py
from slack_bolt import App
import os

# grabs the credentials from .env directly
slack_app = App()
users_map = {}


def send_message(channel: str, thread_ts: str, text: str):
    try:
        slack_app.client.chat_postMessage(
            token=os.environ["SLACK_BOT_TOKEN"],
            channel=channel,
            text=text,
            thread_ts=thread_ts,
        )
    except Exception as e:
        print(e)


def get_thread_messages(channel: str, thread_ts: str):
    try:
        result = slack_app.client.conversations_replies(
            token=os.environ["SLACK_BOT_TOKEN"],
            channel=channel,
            ts=thread_ts,
            include_all_metadata=True,
        )
        return result["messages"]
    except Exception as e:
        print(e)


def get_thread_messages_with_usernames(channel: str, thread_ts: str, bot_id: str):
    thread_messages = get_thread_messages(channel, thread_ts)
    messages_arr = [
        f"HALY: {m['text']}"
        if m.get("bot_id")
        else f"{get_username(m['user'])}: {m['text'].replace(f'<@{bot_id}>', '').strip()}"
        for m in thread_messages
    ]
    return "\n".join(messages_arr)


def find_user_by_id(user_id: str):
    try:
        return slack_app.client.users_info(user=user_id)
    except Exception as e:
        print(e)


def get_username(user_id: str):
    if user_id not in users_map:
        user = find_user_by_id(user_id)
        users_map[user_id] = user["user"]["name"]
    return users_map[user_id]


def command(text: str):
    return "You said: " + text

File: src/services/slackService.ts
import { App } from "@slack/bolt";

const app = new App({
  token: process.env.SLACK_BOT_TOKEN,
  signingSecret: process.env.SLACK_SIGNING_SECRET,
});

(async () => {
  await app.start(3001);
  console.log("⚡️ Haly is running!");
})();

//Doing this locally until we have a DB / redis
const usersMap = new Map<string, string>();

// Publish a message to a channel
// need chat:write scope
export const sendMessage = async (
  channel: string,
  threadTs: string,
  text: string
) => {
  try {
    await app.client.chat.postMessage({
      token: process.env.SLACK_BOT_TOKEN,
      channel,
      text,
      thread_ts: threadTs,
    });
  } catch (error) {
    console.error(error);
  }
};

export const getThreadMessages = async (channel: string, threadTs: string) => {
  try {
    const result = await app.client.conversations.replies({
      token: process.env.SLACK_BOT_TOKEN,
      channel,
      ts: threadTs,
      include_all_metadata: true,
    });
    return result.messages;
  } catch (error) {
    console.error(error);
  }
};

export const getThreadMessagesWithUsernames = async (
  channel: string,
  threadTs: string,
  botId: string
) => {
  const threadMessages = await getThreadMessages(channel, threadTs);
  const messagesArr = await Promise.all(
    threadMessages.map(async (m) => {
      const username = await getUsername(m.user);
      if (m.bot_id) {
        return `Haly: ${m.text}`;
      } else {
        return `${username}: ${m.text.replace(`<@${botId}>`, "").trim()}`;
      }
    })
  );
  return messagesArr;
};

// Call the users.info method using the WebClient
export const findUserById = async (userId: string) => {
  try {
    const result = await app.client.users.info({
      user: userId,
    });

    return result;
  } catch (error) {
    console.error(error);
  }
};

export const getUsername = async (userId: string) => {
  if (!usersMap.get(userId)) {
    const user = await findUserById(userId);
    usersMap.set(userId, user.user.name);
  }
  return usersMap.get(userId);
};

export const command = (text) => {
  return `You said: ${text}`;
};
